src/controllers/PostController.ts

```diff
@@ -0,0 +1,210 @@
+      await Post.findByIdAndUpdate(postid, newPost);
+      return res.json(newPost);
+    } else {
+      return res.json({ Error: "Cannot update another users post." });
+    }
+  }
+
+  async deletePost(req: Request, res: Response) {
+    try {
+      const postid = req.params.postid;
+      const userInformation = req.session;
+
+      //check for valid post id and prevents crash
+      const isValid = idIsValid(postid);
+      if (isValid) {
+        return res.status(422).json({ msg: isValid });
+      }
+
+      //finds user
+      const user = (await User.findOne(
+        { username: userInformation.username },
+        "-password"
+      )) as UserInterface;
+
+      //finds the post
+      const thisPost = (await Post.findById(postid)) as PostInterface;
+
+      //finds the poster
+      const thisPostPoster = await User.findById(
+        thisPost.user,
+        "-password -name -email"
+      );
+
+      //check if post exists and prevents crash from null
+      if (!thisPost || thisPost === null)
+        res.status(404).json({
+          msg: "No post found",
+        });
+
+      //if user role_id === 3 "ADMIN" delete anything he wants
+      if (user.role_id === USER_ROLES.ADM) {
+        //delete post if tests are passed
+        await Post.findByIdAndDelete(postid);
+        return res.status(200).json({
+          msg: "Post deleted successfully",
+        });
+      }
+
+      //if user role_id === 2 "MODERATOR" delete anything but ADMIN posts
+      if (
+        user.role_id === USER_ROLES.MOD &&
+        thisPostPoster.role_id !== USER_ROLES.ADM
+      ) {
+        //delete post if tests are passed
+        await Post.findByIdAndDelete(postid);
+        return res.status(200).json({
+          msg: "Post deleted successfully",
+        });
+      }
+
+      //if user role_id === 1 "USER" and this post was made by the same user
+      if (
+        user.role_id === USER_ROLES.USER &&
+        user._id.toString() === thisPostPoster._id.toString()
+      ) {
+        //delete post if tests are passed
+        await Post.findByIdAndDelete(postid);
+        return res.status(200).json({
+          msg: "Post deleted successfully",
+        });
+      }
+      //check if user is trying to do something he does not have permission to do
+      throw new Error(
+        "Access denied. You do not have permission to delete this post"
+      );
+    } catch (error) {
+      console.log(error);
+      return res.status(500).json({
+        msg: "Error",
+      });
+    }
+  }
+
+  async index(req: Request, res: Response) {
+    try {
+      const posts = (await Post.find({}, "name post -_id")) as PostInterface;
+      res.json(posts);
+    } catch (error) {
+      res.status(500).send(error);
+    }
+  }
+
+  async userPosts(req: Request, res: Response) {
+    try {
+      const id = req.params.id;
+
+      const testId = idIsValid(id);
+      if (testId) {
+        return testId;
+      }
```

Comment:
The return of idIsValid must be true or false and the validation can be done as follows: 
```ts
if(idIsValid(id))
       return `message`;
```